Add enforcement of valid integer range to Number

ximinez opened the pull request:
## High Level Overview of Change

Adds a mechanism to check whether an integer `Number` can be safely and accurately represented..

### Context of Change

`Number` uses an `int64` internally, which has a maximum value approximately `9e18`. However, all of it's precision math is done with values below `1e16`. If an integer value larger than that is converted to `Number`, some of the least-significant digits will be dropped. This is unacceptable for situations that need the exact values that integers provide.

Fortunately, the only features where this imprecision will matter are either unsupported (`SingleAssetVault`) or pending merge (`LendingProtocol`). Before the introduction of these features, only `AMM` used `Number`s, and it used them to calculate rates, qualities, etc. Exact values are not needed for that.

### Type of Change

- [X] Bug fix (non-breaking change which fixes an issue)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Tests (you added tests for code that already exists, or your new feature included in this PR)

## Before / After

* Defines a maximum such that any integer values stored in `Number` must be less than `1e14`. This is 1/100th of the computation range so that certain computed fields can go over this value if necessary.
* Adds the ability to designate a `Number` as holding an integer. This has no functional implications to `Number`, and integer `Numbers` can still store values well beyond the valid range for integers. Transactions and other functionality can check whether a `Number` is valid as part of their validation and transaction processing.
* Once a `Number` is designated as an integer, it can not be undesignated. Any interactions and operations (e.g. assignment, addition) involving any integers will return integer results.
* `XRPAmount`s and `MPTAmount`s converted to `Number`s, including via `STAmount`, will be designated as integers.
* Updates `SingleAssetVault` transactions and operations to check values at the important steps to ensure no precision is lost.

## Future Tasks

Even though this PR is based on `develop`, it is a pre-requisite for #5270, Lending Protocol implementation, which will also need to be updated to use these fields. Other pending PRs that may have consequences from this include #5285, MPT on DEX.

ximinez committed:
Add optional enforcement of valid integer range to Number

ximinez committed:
Make all STNumber fields "soeDEFAULT"

ximinez committed:
Add integer enforcement when converting to XRP/MPTAmount to Number

ximinez committed:
Fix build error - avoid copy

ximinez committed:
Catch up the consequences of Number changes

- Change the Number::maxIntValue to all 9's.
- Add integral() to Asset (copied from Lending)
- Add toNumber() functions to STAmount, MPTAmount, XRPAmount to allow
  explicit conversions with enforcement options.
- Add optional Number::EnforceInteger options to STAmount and STNumber
  ctors, conversions, etc. IOUs are never checked.
- Update Vault transactors, and helper functions, to check restrictions.
- Fix and add Vault tests.

ximinez committed:
fix: Use ".value()" instead of "->" when with STObject::Proxy objects

- Turns out that "Proxy::operator->" is not a safe substitute for
  "Proxy::value()." if the field is not required. The implementation
  is different such that "operator->" will return a null ptr if the
  field is not present. This includes default fields with a value of
  zero!

ximinez committed:
Add a distinction between a "valid" and a "representable" Number

- "valid" means the value is <= Number::maxIntValue, which has been
  changed to maxMantissa / 100. A valid number could get bigger and be
  ok - such as when paying late interest on a loan.
- "representable" means the value is <= Number::maxMantissa. An
  unrepresentable number WILL be rounded or truncated.
- Adds a fourth level of enforcement: "compatible". It is used for
  converting XRP to Number (for AMM), and when doing explicit checks.
- "weak" will now throw if the number is unrepresentable.

ximinez committed:
Merge branch 'develop' into ximinez/lending-number

Bronek commented:
In description, "Before / After" section is missing `compatible` checking level.

Bronek commented on include/xrpl/protocol/STAmount.h:
```diff
@@ -40,6 +40,12 @@ class STAmount final : public STBase, public CountedObject<STAmount>
+    // used during automatic conversions to Number. If not set, the default
+    // behavior is used. It can also be overridden when coverting by using
+    // toNumber().
+    std::optional<Number::EnforceInteger> enforceConversion_;
```
this seems to be the same semantics as simpler:

```c++
Number::EnforceInteger enforceConversion_ = Number::EnforceInteger::none;
```

ximinez replied:
> this seems to be the same semantics as simpler:
> 
> ```c++
> Number::EnforceInteger enforceConversion_ = Number::EnforceInteger::none;
> ```

It is not, because of the default behavior in `STAmount::operator Number()`. If `enforceConversion_` is unseated, then the default behavior is used - XRP becomes `compatible`, MPT becomes `strong`, and IOUs are left alone. If it is set, even if set to `none`, that conversion is used regardless of the asset type. (See also `STAmount::toNumber`.)

You could think of the unseated optional as a fifth level of checking called, `default`, but that doesn't make sense for `Number`.

Bronek commented on include/xrpl/protocol/STAmount.h:
```diff
@@ -135,16 +141,46 @@ class STAmount final : public STBase, public CountedObject<STAmount>
+        enforceConversion_ = enforce;
+        if (!enforce)
+        {
+            // Use the default conversion behavior
```
I don't understand why we need this block. The `operator Number` does not seem to be doing any useful work when `enforceConversion_` is empty (or `none`)

ximinez replied:
> I don't understand why we need this block. The `operator Number` does not seem to be doing any useful work when `enforceConversion_` is empty (or `none`)

If the default conversion uses an option that can throw, and the converted `Number` is invalid, that conversion will throw. Otherwise the result is unneeded. There may be a better way to do it.

Bronek commented on src/test/app/Vault_test.cpp:
```diff
@@ -3597,7 +3597,32 @@ class Vault_test : public beast::unit_test::suite
+                 .amount = d.asset(10)});
+            env(tx, ter{tecPRECISION_LOSS});
+            env.close();
+        });
```
I guess this test means that `vaultMaximumIOUScale` in `Protocol.h` should be cut down, perhaps to 15 ? Which will require futher changes in `Vault_test.cpp`

ximinez replied:
> I guess this test means that `vaultMaximumIOUScale` in `Protocol.h` should be cut down, perhaps to 15 ? Which will require futher changes in `Vault_test.cpp`

Perhaps. The only downside to changing it is that when we have a fuller solution that can represent everything, we'll need to have another value that is amendment-gated. Not a deal breaker, but something to consider. What do you think?

Bronek replied:
I think it's fair to make this change to 15 now, and let future amendments worry about themselves

Bronek reviewed (commented):
This is very extensive change; I wonder if a simpler alternative is possible ?

shawnxie999 commented on include/xrpl/basics/Number.h:
```diff
@@ -13,16 +13,52 @@ class Number;
+     *   than maxMantissa), assignment and other operations will throw.
+     * - strong: Like weak, plus, if the absolute value is invalid (larger than
+     *   maxIntValue), assignment and other operations will throw. This is the
+     *   defalut for automatic conversions from MPTAmount to Number.
```
```suggestion
     *   default for automatic conversions from MPTAmount to Number.
```

shawnxie999 commented on include/xrpl/basics/Number.h:
```diff
@@ -13,16 +13,52 @@ class Number;
+    constexpr static rep maxMantissa = minMantissa * 10 - 1;
+    static_assert(maxMantissa == 9'999'999'999'999'999LL);
+
+    constexpr static rep maxIntValue = maxMantissa / 100;
```
how was this number chosen? could we have chosen a larger value?

ximinez replied:
> how was this number chosen? could we have chosen a larger value?

There's a discussion in the comments in https://ripplelabs.atlassian.net/browse/RIPD-4045.

Since that's not public, the tl;dr is that I started with `maxMantissa / 10`, which would give the `STNumber` fields in `Vault` room to grow if they get, for example, extra interest from a special case payment in a `Loan`. I was concerned that wouldn't be enough overhead. The answer was

> Let’s try to go with a safer, conservative limit as in one that reduces the truncation as much as possible

This is intended to be a temporary solution to get the feature available without a major rewrite. A fuller solution should be available in a version or two which allows the entire range of valid numbers.

ximinez committed:
Merge branch 'develop' into ximinez/lending-number

ximinez commented:
> In description, "Before / After" section is missing `compatible` checking level.

Updated.

ximinez commented:
> This is very extensive change; I wonder if a simpler alternative is possible ?

There might be, but I couldn't think of one.

ximinez committed:
Update include/xrpl/basics/Number.h

Fix typo.

Co-authored-by: Shawn Xie <35279399+shawnxie999@users.noreply.github.com>

ximinez committed:
Merge branch 'develop' into ximinez/lending-number

ximinez commented:
> > This is very extensive change; I wonder if a simpler alternative is possible ?
> 
> There might be, but I couldn't think of one.

I do have one idea for something that would be simpler to implement and reason about, but with one significant caveat.

1. Convert `Number` to use `boost::multiprecision::uint128_t` for the mantissa. This is a class we have used without problems in other places, so it's a known quantity.
2. Abstract the `minMantissa`, `maxMantissa` values behind a lookup function. Say, `getMantissaRange`.
3. `getMantissaRange` will check if either the SAV or LP amendments are enabled. If they are not, return the current 16-digit values. If one or both are, return 19-digit values that will include the entire range of valid 63-bit integers.
4. Converting `STAmount` to `Number` will basically be transparent.
5. Converting `Number` to `STAmount` will require scaling the mantissa down with proper rounding to fit into the `STAmount` range, if the `STAmount` is non-integer. i.e. IOU. If it is integer, it will have to check that the value fits in a `uint64_t`.
6. Serializing an `STNumber` will have to do similar scaling as an `STAmount`, but bringing everything to fit within a 63- or 64-bit mantissa, also with proper rounding.
7. Deserializing an `STNumber` will be trivial, because they values will just fit.

*The caveat is step 3* - `Number` lives in `basics`, and `Rules` lives in `protocol` (*oh, and so do features*). We'd have to find a way to communicate the amendment status without breaking levelization.

AMM behavior may change slightly because of the extra precision when one of the amendments activates, but by definition, it will be a consistent, amendment-gated change across all nodes. We can get an idea of how disruptive that would be by seeing how existing tests break.

But now we've got no extra state in `Number`, no changes to transactors, and minimal changes to `Number`, `STNumber`, and `STAmount`.

What do you think?

gregtatcam commented:
> > > This is very extensive change; I wonder if a simpler alternative is possible ?
> > 
> > 
> > There might be, but I couldn't think of one.
> 
> I do have one idea for something that would be simpler to implement and reason about, but with one significant caveat.
> 
> 1. Convert `Number` to use `boost::multiprecision::uint128_t` for the mantissa. This is a class we have used without problems in other places, so it's a known quantity.
> 2. Abstract the `minMantissa`, `maxMantissa` values behind a lookup function. Say, `getMantissaRange`.
> 3. `getMantissaRange` will check if either the SAV or LP amendments are enabled. If they are not, return the current 16-digit values. If one or both are, return 19-digit values that will include the entire range of valid 63-bit integers.
> 4. Converting `STAmount` to `Number` will basically be transparent.
> 5. Converting `Number` to `STAmount` will require scaling the mantissa down with proper rounding to fit into the `STAmount` range, if the `STAmount` is non-integer. i.e. IOU. If it is integer, it will have to check that the value fits in a `uint64_t`.
> 6. Serializing an `STNumber` will have to do similar scaling as an `STAmount`, but bringing everything to fit within a 63- or 64-bit mantissa, also with proper rounding.
> 7. Deserializing an `STNumber` will be trivial, because they values will just fit.
> 
> _The caveat is step 3_ - `Number` lives in `basics`, and `Rules` lives in `protocol` (_oh, and so do features_). We'd have to find a way to communicate the amendment status without breaking levelization.
> 
> AMM behavior may change slightly because of the extra precision when one of the amendments activates, but by definition, it will be a consistent, amendment-gated change across all nodes. We can get an idea of how disruptive that would be by seeing how existing tests break.
> 
> But now we've got no extra state in `Number`, no changes to transactors, and minimal changes to `Number`, `STNumber`, and `STAmount`.
> 
> What do you think?

Sounds good. Love it! One question though. Would there be any performance impact because of 128-bit?

Bronek commented:
> _The caveat is step 3_ - `Number` lives in `basics`, and `Rules` lives in `protocol` (_oh, and so do features_). We'd have to find a way to communicate the amendment status without breaking levelization.

That's actually not so difficult IMO. You revert the dependency here by:
* next to `Number` define an enum to be used as a parameter in `getMantissaRange` 
* unit tests for `Number` just use this enum
* define a helper function somewhere in `protocol` layer to return that enum value, and taking `Rules const&` as a parameter
* voila. Every user of `Number` *typically* has access to `Rules` object, they just need to remember to pass it to  the helper function when calling `Number` functions

I see other difficulty though - **which** `Number` functions would need  to read that mantissa range ? You cannot pass an extra parameter to overloaded operators.

shawnxie999 commented:
What's the guideline for when/where each kind of enforcement type should be used? Is it `weak` for IOU and `strong` for MPT?

ximinez commented:
> Sounds good. Love it! One question though. Would there be any performance impact because of 128-bit?

We'll have to do some testing to see if there is or not. I'm really hoping not.

ximinez commented:
> What's the guideline for when/where each kind of enforcement type should be used? Is it `weak` for IOU and `strong` for MPT?

In this current PR, the defaults are
* `none` for IOUs
* `compatible` for XRP
* `strong` for MPT

In places where it's set manually, it depends on the particular context.

shawnxie999 commented:
> > What's the guideline for when/where each kind of enforcement type should be used? Is it `weak` for IOU and `strong` for MPT?
> 
> In this current PR, the defaults are
> 
> * `none` for IOUs
> * `compatible` for XRP
> * `strong` for MPT
> 
> In places where it's set manually, it depends on the particular context.

how about `weak`? i see that being used in several places in this PR

ximinez committed:
Merge remote-tracking branch 'XRPLF/develop' into ximinez/lending-number

* XRPLF/develop:
  chore: Set version 3.1.0-b0 (5986)
  ci: Clean workspace on Windows self-hosted runners (6024)
  docs: fix spelling in comments (6002)
  fix: floating point representation errors in vault (5997)
  ci: Specify bash as the default shell in workflows (6021)
  refactor: Add `XRPL_RETIRE_FIX` and `XRPL_RETIRE_FEATURE` macros (6014)
  refactor: Retire DepositPreAuth and DepositAuth amendments (5978)
  chore: Move running of unit tests out of coverage target (6018)
  refactor: Retire PayChanRecipientOwnerDir amendment (5946)

ximinez committed:
Merge branch 'develop' into ximinez/lending-number

ximinez committed:
Merge branch 'develop' into ximinez/lending-number

ximinez committed:
Rip out about half the code: levels, enforcement, and STAmount changes

- Changed the EnforceInteger enum into a bool.
- Removed enforcement by throw.
- Essentially got rid of the "weak" and "strong" options.
- Removed integer options from STAmount. Since there's no throwing,
  there's no need to override the default.

ximinez committed:
fixup! Rip out about half the code: levels, enforcement, and STAmount changes

ximinez committed:
Fixes build errors and test failures

- STAmount::clear() needs to use 0 exponent for all integral types.
- Change vaultMaximumIOUScale to 13.
- Fix an dereferenced unseated optional.

gregtatcam commented:
Some Vault unit-tests should be updated because `vaultMaximumIOUScale` has been changed from 18 to 13.

gregtatcam commented on src/libxrpl/basics/Number.cpp:
```diff
@@ -207,9 +207,52 @@ Number::normalize()
 {
+    // The strictest setting prevails
+    if (!isInteger_)
+        isInteger_ = y.isInteger_;
```
This doesn't seem right. Consider this test:

        Number i{100, true};
        Number a{1, -10, false};
        Number b = a + i;
        std::cout << i << " " << a << " " << b << " " << b.isInteger() << std::endl;

It outputs:

    100 0.0000000001 100.0000000001 1

'b' is integer but it's actually not.

ximinez replied:
> This doesn't seem right. Consider this test:
> 
> [...]
> 
> 'b' is integer but it's actually not.

This is completely intentional, and is why there are no asserts or throws related to the flag. 

What I wrote in Number.h is
```
    // isInteger_ is not serialized, transmitted, or used in
    // calculations in any way. It is used only for internal validation
    // of integer types. It is a one-way switch. Once it's on, it stays
    // on.
```
I should probably also add that it's transmissible, or "sticky", so that any interaction with an integer will become an integer. The only thing that can strip the integer flag from a `Number` is converting or assigning it to something other than a `Number`.

The reason is that the flag does not affect any operations, and can be completely ignored if it's not needed. If it is needed, it would be much worse to strip it off than it would be to add it unnecessarily.

To get a better view of what I'm talking about, it's probably better to look at how the flag is used: The `Number::valid()` and `Number::representable()` functions, and the `STAmount::validNumber()` and `STAmount::representableNumber()` functions which wrap them.

`git grep -w -e "representable()" -e "valid()" -e "validNumber()" -e "representableNumber()"`

Notice how they are only called from tests, and specific transactors doing specific tests.

ximinez committed:
fix: Number: Do not attempt to take the negative of an unsigned

ximinez committed:
test fix: Remove cases with Vault scale 18

- The limit has been reduced to 13, so 18 is now considered malformed.
- Make the test more robust so it doesn't segfault if the vault creation
  fails.

ximinez commented:
> Some Vault unit-tests should be updated because `vaultMaximumIOUScale` has been changed from 18 to 13.

Two things:
1. I changed `vaultMaximumIOUScale` to 15, based on @Bronek's earlier suggestion, and the fact that 15 is representable.
2. I'm getting ready to push a commit that will update the relevant unit tests.

ximinez committed:
Change the vaultMaximumIOUScale from 13 to 15

- Anything above 13 is _nearly_ unusable, but there might be some edge
  use cases where it is.
- Added unit tests to show this.

ximinez committed:
Expand the description of Number::isInteger_

gregtatcam commented on include/xrpl/basics/Number.h:
```diff
@@ -35,15 +56,42 @@ class Number
+    bool
+    isInteger() const noexcept;
+
+    bool
```
Could you please add function description to `valid` and `representable`?

ximinez replied:
Done. I also renamed `valid()` to `fits()`

ximinez committed:
Merge branch 'develop' into ximinez/lending-number

ximinez committed:
Rewrite to clarify Number limitations, enforce limits in conversion

- Overflow Number -> int64 if it doesn't fit in the mantissa range.
- Only enabled if at least one of the SingleAssetVault or
  LendingProtocol amendments are enabled.
- Will throw the overflow error if the value is larger than maxMantissa.
  Current behavior is to throw if the value is larger than max int64_t
  value.

ximinez committed:
Avoid some null dereferences in tests

ximinez committed:
Add a couple of high value MPT tests

ximinez committed:
Add an RAII NumberIntegerOverflowGuard class, and update tests

ximinez committed:
Fix large integer overflow

- Document why STAmount to Number conversions work in canonicalize().
- Update some unit tests

ximinez committed:
Fix Vault tests due to Number changes

ximinez committed:
More test and logging updates

ximinez committed:
Set the global rules for every transaction step

- Amendment gated on SAV and LP

ximinez committed:
Merge branch 'develop' into ximinez/lending-number

ximinez committed:
Merge branch 'develop' into ximinez/lending-number

ximinez committed:
Merge branch 'develop' into ximinez/lending-number

ximinez committed:
Merge branch 'develop' into ximinez/lending-number

ximinez committed:
Merge branch 'develop' into ximinez/lending-number

ximinez committed:
Merge branch 'develop' into ximinez/lending-number
